larsim/EventGenerator/MultiPart/MultiPartVertex_module.cc

```diff
@@ -0,0 +1,762 @@
+  double phi = fFlatRandom->fire(0.0, 2.0 * M_PI);
+  double px = st * TMath::Cos(phi);
+  double py = st * TMath::Sin(phi);
+  double pz = ct;
+  std::array<double, 3U> result = {px, py, pz};
+  return result;
+}
+/*
+std::array<double, 3U> MultiPartVertex::extractDirection() const {
+    double px, py, pz;
+    px = fNormalRandom->fire(0, 1);
+    py = fNormalRandom->fire(0, 1);
+    pz = fNormalRandom->fire(0, 1);
+    double p = std::hypot(px, py, pz);
+    px = px / p;
+    py = py / p;
+    pz = pz / p;
+    std::array<double, 3U> result = { px, py, pz };
+    return result;
+}
+*/
+
+void MultiPartVertex::GenMomentum(const PartGenParam& param,
+                                  const double& mass,
+                                  double& px,
+                                  double& py,
+                                  double& pz)
+{
+
+  double tot_energy = 0;
+  if (param.use_mom)
+    tot_energy =
+      sqrt(cet::square(fFlatRandom->fire(param.kerange[0], param.kerange[1])) + cet::square(mass));
+  else
+    tot_energy = fFlatRandom->fire(param.kerange[0], param.kerange[1]) + mass;
+
+  double mom_mag = sqrt(cet::square(tot_energy) - cet::square(mass));
+
+  std::array<double, 3U> p = extractDirection();
+  px = p[0];
+  py = p[1];
+  pz = p[2];
+
+  if (_debug > 1)
+    std::cout << "    Direction : (" << px << "," << py << "," << pz << ")" << std::endl
+              << "    Momentum  : " << mom_mag << " [MeV/c]" << std::endl
+              << "    Energy    : " << tot_energy << " [MeV/c^2]" << std::endl;
+  px *= mom_mag;
+  py *= mom_mag;
+  pz *= mom_mag;
+}
+
+void MultiPartVertex::GenMomentum(const PartGenParam& param,
+                                  const double& mass,
+                                  double& px,
+                                  double& py,
+                                  double& pz,
+                                  bool& same_range)
+{
+
+  if (!same_range) { GenMomentum(param, mass, px, py, pz); }
+  else {
+    double tot_energy = 0;
+    if (param.use_mom)
+      tot_energy = sqrt(cet::square(fFlatRandom->fire(param.kerange[0], param.kerange[1])) +
+                        cet::square(mass));
+    else
+      tot_energy = fFlatRandom->fire(param.kerange[0], 1) + mass;
+
+    double mom_mag = sqrt(cet::square(tot_energy) - cet::square(mass));
+
+    std::array<double, 3U> p = extractDirection();
+    px = p[0];
+    py = p[1];
+    pz = p[2];
+
+    if (_debug > 1)
+      std::cout << "    Direction : (" << px << "," << py << "," << pz << ")" << std::endl
+                << "    Momentum  : " << mom_mag << " [MeV/c]" << std::endl
+                << "    Energy    : " << tot_energy << " [MeV/c^2]" << std::endl;
+    px *= mom_mag;
+    py *= mom_mag;
+    pz *= mom_mag;
+  }
+}
+
+void MultiPartVertex::GenMomentumSF(const double& sf,
+                                    const double& m,
+                                    const double& p,
+                                    double& p_sf)
```

Comment:
Suggest altering the function signature so that `p_sf` is the returned value:

```c++
double MultiPartVertex::GenMomentumSF(double sf, double m, double p);
```